src/_P039_Thermosensors.ino

```diff
@@ -0,0 +1,1681 @@
+  // as possible.
+  const bool Plugin_039_SensorAttached = !(rawvalue & 0x0004);
+
+  if (Plugin_039_SensorAttached)
+  {
+    // Shift RAW value 3 Bits to the right to get the data
+    rawvalue >>= 3;
+
+    // Calculate Celsius
+    return rawvalue * 0.25f;
+  }
+  else
+  {
+    return NAN;
+  }
+}
+
+float readMax31855(struct EventStruct *event)
+{
+  union {
+      uint8_t messageBuffer[4] = {0x00, 0x00, 0x00, 0x00};
+      uint32_t value;
+  } u_mB;
+
+  uint8_t CS_pin_no = get_SPI_CS_Pin(event);
+
+  // u_mB.messageBuffer[4] = {0x00, 0x00, 0x00, 0x00};
+
+  // "transfer" 0x0 and read the 32 Bit conversion register from the Chip
+
+  transfer_n_ByteSPI(CS_pin_no, 4, &u_mB.messageBuffer[0]);
+  uint16_t rawvalue = u_mB.value;
+
+  // uint32_t rawvalue = read_n_ByteitRegister(CS_pin_no, 0x0000);
+
+  // // Shift MSB 16 Bits to the left
+  // rawvalue <<= 16;
+
+  // // "transfer" 0x0 and read the LSB Data from the Chip
+  // rawvalue |= read16BitRegister(CS_pin_no, 0x0000);
+
+  # ifndef BUILD_NO_DEBUG
+
+    if (loglevelActiveFor(LOG_LEVEL_DEBUG)) 
+    {
+      String log;
+      if((log.reserve(66u))) {
+        log = F("P039 : MAX31855 : RAW - BIN:");
+        log += String(rawvalue, BIN);
+        log += F(" rawvalue,HEX: ");
+        log += String(rawvalue, HEX);
+        log += F(" rawvalue,DEC: ");
+        log += String(rawvalue);
+        log += F(" u_mB.value,DEC: ");
+        log += String(u_mB.value, DEC);
+        log += F(" u_mB.messageBuffer[],DEC:");
+        for (size_t i = 0u; i < 4; i++)
+        {
+                log += ' ';
+                log += String(u_mB.messageBuffer[i], DEC);
+        }
+        addLog(LOG_LEVEL_DEBUG, log);
+      }
+    }
+
+  # endif // ifndef BUILD_NO_DEBUG
+
+
+  // FIXME TD-er: This static flag is shared among all instances of this plugin
+  static bool sensorFault = false;
+
+  if (sensorFault != ((rawvalue & 0x7) == 0)) {
+    // Fault code changed, log them
+    sensorFault = ((rawvalue & 0x7) == 0);
+
+    # ifndef BUILD_NO_DEBUG
+
+      if (loglevelActiveFor(LOG_LEVEL_DEBUG)) 
+      {
+        String log;
+        if((log.reserve(66u))) {
+          log = F("P039 : MAX31855");
+
+          if (!sensorFault) {
+            log += F("Fault resolved");
+          } else {
+            log += F("Fault code:");
+
+            if (rawvalue & 0x01) {
+              log += F(" Open (no connection)");
+            }
+
+            if (rawvalue & 0x02) {
+              log += F(" Short-circuit to GND");
+            }
+
+            if (rawvalue & 0x04) {
+              log += F(" Short-circuit to Vcc");
+            }
+          }
+          addLog(LOG_LEVEL_DEBUG, log);
+        }
+      } 
+
+    # endif // ifndef BUILD_NO_DEBUG
+
+  }
+
+  // D16 - This bit reads at 1 when any of the SCV, SCG, or OC faults are active. Default value is 0.
+  const bool Plugin_039_SensorAttached = !(rawvalue & 0x00010000);
```

Comment:
@chri-kai-in @TD-er D17,D3 are always 0, we should check if D16 and D2,D1,D0 are also 0, so we are sure about sensor here. My test shows when one of faults is 1, conversion value is unpredictable.

Reply:
Next commit/push will take some time, as i need to dig into the thermocouple devices and the TIMER_API. I will gather some of your next review findings - keep on reviewing and commenting !

Reply:
@chri-kai-in got hint for you in other comment

Reply:
Btw. i checked data sheet again - it states D16 goes high if any of the detailed failure flags is active . Checking D16 only should cover all cases.  Anyhow: I will add the other bits to go for a safe programming approach, "just in case" hardware would behave not according to its datasheet.-> does make sense, would not be the first time...